shdlc/sensirion_shdlc.c

```diff
@@ -201,3 +201,184 @@ int16_t sensirion_shdlc_rx(uint8_t max_data_len,
+            return;
+    }
+}
+
+void sensirion_shdlc_add_uint8_t_to_frame(
+    struct sensirion_shdlc_buffer* tx_frame, uint8_t data) {
+
+    sensirion_shdlc_stuff_byte(tx_frame, data);
+    tx_frame->checksum += data;
+}
+
+void sensirion_shdlc_begin_frame(struct sensirion_shdlc_buffer* tx_frame,
+                                 uint8_t command, uint8_t address,
+                                 uint8_t data_length) {
+    tx_frame->checksum = 0;
+    tx_frame->offset = 0;
+    tx_frame->data[tx_frame->offset++] = SHDLC_START;
+    sensirion_shdlc_add_uint8_t_to_frame(tx_frame, address);
+    sensirion_shdlc_add_uint8_t_to_frame(tx_frame, command);
+    sensirion_shdlc_add_uint8_t_to_frame(tx_frame, data_length);
+}
+
+void sensirion_shdlc_add_uint32_t_to_frame(
+    struct sensirion_shdlc_buffer* tx_frame, uint32_t data) {
+
+    sensirion_shdlc_add_uint8_t_to_frame(tx_frame,
+                                         (uint8_t)((data & 0xFF000000) >> 24));
+    sensirion_shdlc_add_uint8_t_to_frame(tx_frame,
+                                         (uint8_t)((data & 0x00FF0000) >> 16));
+    sensirion_shdlc_add_uint8_t_to_frame(tx_frame,
+                                         (uint8_t)((data & 0x0000FF00) >> 8));
+    sensirion_shdlc_add_uint8_t_to_frame(tx_frame,
+                                         (uint8_t)((data & 0x000000FF) >> 0));
+}
+
+void sensirion_shdlc_add_int32_t_to_frame(
+    struct sensirion_shdlc_buffer* tx_frame, int32_t data) {
+
+    sensirion_shdlc_add_uint32_t_to_frame(tx_frame, (uint32_t)data);
+}
+
+void sensirion_shdlc_add_uint16_t_to_frame(
+    struct sensirion_shdlc_buffer* tx_frame, uint16_t data) {
+
+    sensirion_shdlc_add_uint8_t_to_frame(tx_frame,
+                                         (uint8_t)((data & 0xFF00) >> 8));
+    sensirion_shdlc_add_uint8_t_to_frame(tx_frame,
+                                         (uint8_t)((data & 0x00FF) >> 0));
+}
+
+void sensirion_shdlc_add_int16_t_to_frame(
+    struct sensirion_shdlc_buffer* tx_frame, int16_t data) {
+
+    sensirion_shdlc_add_uint16_t_to_frame(tx_frame, (uint16_t)data);
+}
+
+void sensirion_shdlc_add_float_to_frame(struct sensirion_shdlc_buffer* tx_frame,
+                                        float data) {
+    union {
+        uint32_t uint32_data;
+        float float_data;
+    } convert;
+
+    convert.float_data = data;
+    sensirion_shdlc_add_uint32_t_to_frame(tx_frame, convert.uint32_data);
+}
+
+void sensirion_shdlc_add_bytes_to_frame(struct sensirion_shdlc_buffer* tx_frame,
+                                        uint8_t* data, uint16_t data_length) {
+    uint16_t i;
+
+    for (i = 0; i < data_length; i++) {
+        sensirion_shdlc_add_uint8_t_to_frame(tx_frame, data[i]);
+    }
+}
+
+int16_t sensirion_shdlc_tx_frame(struct sensirion_shdlc_buffer* tx_frame) {
+
+    int16_t tx_length;
+
+    sensirion_shdlc_add_uint8_t_to_frame(tx_frame, ~(tx_frame->checksum));
+    tx_frame->data[tx_frame->offset++] = SHDLC_STOP;
```

Comment:
Can we extract this into a `sensirion_shdlc_finish_frame()`? So we can create a complete frame without needing to call into the HAL.